### PR TITLE
Allow initializing a chain of protos using only a hash in Ruby

### PR DESCRIPTION
**Pull request by zanker**
I saw #3120  also asked for this. We're in the process of trying to migrate people from a pure Ruby proto gem to the Google one, and the requirement to initialize all protos is a blocker. Sounds like there was interest in this, so figured I'd take a stab at implementing it in.

Taking the examples from #3120, this lets you do: `MyMessage.new(submessage_field: { foo: 42 })`and it will automatically turn the `{ foo: 42 }` into a `SubMessage.new(foo: 42)`.

Since a common scenario is also to be turning JSON into protos (for proto backed JSON APIs), I also changed the initialization to support symbol or string keys to prevent having to do a deep symbolization pass before initializing the protos.

### CODE REVIEW

**Comment by bazel-io** [comment]
Can one of the admins verify this patch?

**Comment by haberman** [comment]
ok to test

**Comment by haberman** [comment]
Looks like this change failed Travis?

**Comment by zanker** [comment]
@haberman Fixed the test that failed. It was one of the compatibility tests, which now fails with `RangeError` because it's now checking if `z` is an enum value, whereas previously it would fail with `TypeError` since it was a string.

Is the fact that the exception changed ok?

File: ruby/ext/google/protobuf_c/message.c [hunk 1]
```diff
@@ -248,16 +260,26 @@ int Message_initialize_kwarg(VALUE key, VALUE val, VALUE _self) {
     Map_merge_into_self(map, val);
   } else if (upb_fielddef_label(f) == UPB_LABEL_REPEATED) {
     VALUE ary;
+    VALUE entry;
```
Comment (haberman) [inline ruby/ext/google/protobuf_c/message.c]: This is only used inside the for loop, so please declare it there.

Comment (zanker) [reply ruby/ext/google/protobuf_c/message.c]: Done!

**Comment by TeBoring** [comment]
@zanker could you take a look for #4809?

**Pull request merged**